bin/cron/build_contact_rollups

```diff
@@ -6,47 +6,56 @@ require 'cdo/contact_rollups_validation'
 require 'cdo/pardot'
 require 'cdo/chat_client'
 require 'cdo/only_one'
+require 'cdo/log_collector'
 require 'aws-sdk-rds'
+require 'honeybadger/ruby'
 
 CHAT_CHANNEL = "cron-daily".freeze
 
 def main
-  time_start = Time.now
+  start_time = Time.now
+  log_collector = LogCollector.new
 
   # Delete clone database cluster if it was left behind by the last execution of Contact Rollups.
-  ContactRollups.delete_database_clone
-  ContactRollups.create_database_clone
+  log_collector.time!('Delete db clone') {ContactRollups.delete_database_clone}
+  log_collector.time!('Create db clone') {ContactRollups.create_database_clone}
 
   # Build new daily table of contact rollups
-  ContactRollups.build_contact_rollups
+  log_collector.time!('Build contact rollups') {ContactRollups.build_contact_rollups(log_collector)}
 
   # Validate that daily contact rollups meets sanity expectations
-  results = ContactRollupsValidation.validate_contact_rollups
-  unless results[:pass]
-    message = "Contact rollups process failed. "\
-              "Validation of daily rollup failed:\n#{results[:output]}"
-    ChatClient.message(CHAT_CHANNEL, message)
-    # Raise exception so this goes to HoneyBadger
-    raise message
+  log_collector.time!('Validate contact rollups') do
+    validation_results = ContactRollupsValidation.validate_contact_rollups
+
+    raise "Validation of daily rollup failed:\n#{validation_results[:output]}" unless validation_results[:pass]
   end
 
   # Sync daily rollup to master rollup and compute deltas
-  ContactRollups.sync_contact_rollups_to_main
+  log_collector.time!('Sync to main db') {ContactRollups.sync_contact_rollups_to_main(log_collector)}
 
   # Sync deltas into Pardot
-  results = Pardot.sync_contact_rollups_to_pardot
+  log_collector.time!('Sync to Pardot') do
+    sync_results = Pardot.sync_contact_rollups_to_pardot
+    log_collector.info(
+      "Syncing completed: #{sync_results[:num_inserts]} inserts, #{sync_results[:num_updates]} updates."
+    )
+  end
+rescue => e
+  log_collector.record_exception(e)
+ensure
+  log_collector.time('Clean up db clone') {ContactRollups.delete_database_clone}
+  log_collector.info("Script finished with #{log_collector.errors.size} caught exception(s) in"\
+    " #{LogCollector.get_friendly_time(Time.now - start_time)}."
+  )
 
-  time_elapsed = (Time.now - time_start).to_i
-  message = "Contact rollups process complete. #{results[:num_inserts]} "\
-    "inserts and #{results[:num_updates]} updates to Pardot. "\
-    "Total time: #{time_elapsed} sec"
+  ChatClient.message(CHAT_CHANNEL, log_collector.to_s)
 
-  ChatClient.message(CHAT_CHANNEL, message)
+  # Notify HoneyBadger about the last caught exception.
```

Comment:
why do we only care about the last exception?

Reply:
I'm still debating about this, it takes only 1 fatal exception to stop the process and it would be the last caught exception. Other exceptions are non-fatal. However, tasks in the `ensure` block could also raise exception, making the previous assumption invalid.

Will update comment to explain the decision.